### PR TITLE
stats: Remove field from GaugeImpl that was not used, by moving it to the counter-specific section.

### LINKED ISSUE #7476: stats: potential race between Gauge mergeImportMode and value-setting
During debug of an earlier issue, @mattklein123 and I stumbled on a potential race for Gauges, where we may race between Gauge::mergeImportMode, which can clear `value_`, and any of the value_-updating methods.

### PR DESCRIPTION
**Pull request by jmarantz**
Description: forks more of the implementations between Counters and Gauges. Counters remain effectively as they were. For Gauges, we get to drop the `pending_increment_` field which was not used, but to resolve this bug we need to add a mutex, so this saves a little memory.

The need for this is mentioned in #7109 but this does not fully resolve that bug.

This is WiP because I need to verify that it passes 'release' test now, having tweaked the expected memory value, but no way to check it on my system.
Risk Level: medium
Testing: //test/common/stats/...
Docs Changes: n/a
Release Notes: n/a

### CODE REVIEW

**Comment by mattklein123** [comment]
I've thought about this since we originally talked about this, and I don't think I'm in favor of adding a lock to all gauge operations. I could see this easily becoming contended on high core count servers.

Can we spec out what a solution would look like that does not require a lock inside the gauge? Last time we talked I think we discussed locking just the creation operation, which would be more complicated as it would involve the store, but would not require a lock around all operations? WDYT?

/wait-any

**Comment by jmarantz** [comment]
I hear what you're saying; and why add burden across the board for a very rare event.

I do have a question about contention in practice and atomics. If you look up the documentation on std::atomic with the default memory-ordering (std::memory_order_seq_cst), which is what we use, my reading is it's got the ordering and contention properties of a mutex lock for each update. What we were doing in this PR is taking a single lock to update 2 variables rather than taking a lock for each of the 2 variables. It doesn't feel that it's likely to be different in performance to me, but I admit I haven't benchmarked the difference. I think you have a much stronger feel for the performance details of this than I do.

In any case, I am re-purposing this PR to simply remove the dead field from the gauge structure, and make it counter-specific.

**Comment by mattklein123** [comment]
> I do have a question about contention in practice and atomics. If you look up the documentation on std::atomic with the default memory-ordering (std::memory_order_seq_cst), which is what we use, my reading is it's got the ordering and contention properties of a mutex lock for each update. What we were doing in this PR is taking a single lock to update 2 variables rather than taking a lock for each of the 2 variables. It doesn't feel that it's likely to be different in performance to me, but I admit I haven't benchmarked the difference. I think you have a much stronger feel for the performance details of this than I do.

I'm not sure about the implementation of absl::Mutex, but in a best case non-contended scenario it would require 2 atomic operations to lock and unlock. In a contended scenario, it may spin, use ticket queues, ultimately sleep, etc. So, yeah, in a non-contended state I think the performance will be similar, though without a doubt the mutex will require more supporting instructions. In a contended state, the raw atomics will perform a lot better, but you are correct that they too will fall apart at some point. See https://github.com/envoyproxy/envoy/issues/509. :) (But my preference would be to try to be as lightweight as possible here to avoid falling apart for as long as possible...)

File: test/integration/stats_integration_test.cc [hunk 1]
```diff
@@ -209,6 +209,7 @@ TEST_P(ClusterMemoryTestRunner, MemoryLargeClusterSizeWithStats) {
   // 2019/06/17  7243     49412       49700   macros for exact/upper-bound memory checks
   // 2019/06/29  7364     45685       46000   combine 2 levels of stat ref-counting into 1
   // 2019/06/30  7428     42742       43000   remove stats multiple inheritance, inline HeapStatData
+  // 2019/07/06  7477     42742       43000   fork gauge representation to drop pending_increment_
```
Comment (mattklein123) [inline test/integration/stats_integration_test.cc]: Q: Wouldn't you expect this to drop the size? Or is this lost in padding somewhere?

Comment (jmarantz) [reply test/integration/stats_integration_test.cc]: Yes I did expect that. I confirmed in the debugger that sizeof(GaugeImpl) is now 40 whereas it used to be 48. I have two theories:

1. tcmalloc uses a common size bucket for 40-byte objects and 48-byte objects.
2. there are not enough Gauges per cluster to impact the overall allocation size per cluster.

I could easily confirm these theories, except my workstation crashed with a kernel panic and won't come up now :). I am currently waiting for the 4 winds to come and make that problem go away. Or I may just bring up an environment in a VM to see what's going on.

Comment (mattklein123) [reply test/integration/stats_integration_test.cc]: My guess is tcmalloc allocates a 64-byte chunk. It's not a big deal if you want to investigate later. I will approve so you can decide.

**Comment by jmarantz** [comment]
Thanks Matt -- I'll go ahead and merge as I will not be able to resolve any master-merge needs in this PR anytime soon :)

**Pull request merged**